### PR TITLE
feat: Add blocking timeout in polling config manager.

### PR DESCRIPTION
**Pull request by yasirfolio3**
## Summary
This introduces blocking timeout in PollingConfigManager. Config Manager will block any calling go-routine until the config is available or the timeout occurs.

### CODE REVIEW

**Comment by msohailhussain** [comment]
converage is decreased, @yasirfolio3 let's discuss how coverage can be increased.

File: pkg/utils/channel_utils.go [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+ * limitations under the License.                                           *
+ ***************************************************************************/
+
+// Package utils //
+package utils
+
+import "time"
+
+// IsChannelClosed will check if the given channel is closed
+func IsChannelClosed(ch chan struct{}) bool {
```
Comment (mjc1283) [inline pkg/utils/channel_utils.go]: Nit: can we use read-only channel type `<-chan struct{}`? And also below in `WaitForChannelToCloseOrTimeout`?

File: pkg/config/polling_manager_test.go [hunk 2]
```diff
@@ -146,6 +150,127 @@ func TestNewAsyncPollingProjectConfigManagerWithNullDatafile(t *testing.T) {
...
+
+	configManager := NewPollingProjectConfigManager(sdkKey, WithRequester(mockRequester), WithBlockingTimeout(blockingTimeout))
+	mockRequester.AssertExpectations(t)
+
+	start := time.Now()
+	expectedExecutionTime := start.Add(blockingTimeout)
+	_, _ = configManager.GetConfig()
+	actualExecutionTime := start.Add(time.Since(start))
+	assert.WithinDuration(t, expectedExecutionTime, actualExecutionTime, 10*time.Millisecond)
+}
```
Comment (mjc1283) [inline pkg/config/polling_manager_test.go]: This test case and the others with 50ms blocking timeout seem to have the potential to be flaky, and will run very slowly. I will think about other ways to test this behavior - @pawels-optimizely @mikeng13 @mikecdavis any ideas?

**Comment by yasirfolio3** [comment]
> Nit in PR description:
> 
> > Config Manager will block the main thread until the config is available or the timeout occurs
> 
> What is the "main thread"? It will block any calling goroutine, right?

correct, thanks for the correction.

File: pkg/config/polling_manager_test.go [hunk 3]
```diff
@@ -146,6 +150,129 @@ func TestNewAsyncPollingProjectConfigManagerWithNullDatafile(t *testing.T) {
...
+
+	configManager := NewPollingProjectConfigManager(sdkKey, WithRequester(mockRequester), WithBlockingTimeout(blockingTimeout))
+	mockRequester.AssertExpectations(t)
+
+	start := time.Now()
+	expectedExecutionTime := start.Add(blockingTimeout)
+	// blocking api for specified time.
+	_, _ = configManager.GetConfig()
+	actualExecutionTime := start.Add(time.Since(start))
+	assert.WithinDuration(t, expectedExecutionTime, actualExecutionTime, 10*time.Millisecond)
```
Comment (mjc1283) [inline pkg/config/polling_manager_test.go]: @yasirfolio3 @msohailhussain Following up on my earlier comment.

To limit flakiness I suggest we remove this `assert.WithinDuration`. Instead, we can implement a timeout on the total execution time of this test case, failing when it exceeds a pretty large duration, like 1-2 seconds. This lets us be resilient to varying execution time of `GetConfig`, while still ensuring that our test fails when the implementation of the timeout inside `GetConfig` is broken.

File: pkg/config/polling_manager_test.go [hunk 4]
```diff
@@ -146,6 +150,129 @@ func TestNewAsyncPollingProjectConfigManagerWithNullDatafile(t *testing.T) {
 	mockRequester.AssertExpectations(t)
 }
 
+func TestGetConfigForPollingConfigManagerWaitsForTimeoutForInvalidDatafile(t *testing.T) {
+	blockingTimeout := 50 * time.Millisecond
```
Comment (mjc1283) [inline pkg/config/polling_manager_test.go]: I suggest reducing this to 1 millisecond, combined with my other suggestion, there is no need for this to be so high.

**Pull request closed without merging**